src/env.rs

```diff
@@ -0,0 +1,296 @@
+/// this way.
+pub trait ValueType<'a>: Sized {
+    /// The corresponding owned type.
+    type Owned;
+
+    /// Attempt to convert the generic `Value` into this type.
+    fn try_from_value(v: &'a Value) -> Result<Self, String>;
+}
+
+impl Env {
+    /// Gets a value from the environment, expecting it to be present.
+    ///
+    /// Note that the return value is a reference for "expensive" types such
+    /// as strings, but an ordinary value for "cheap" types such as numbers
+    /// and colors.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the key is not found, or if it is present with the wrong type.
+    pub fn get<'a, V: ValueType<'a>>(&'a self, key: Key<V>) -> V {
+        if let Some(value) = self.0.map.get(key.key) {
+            value.to_inner_unchecked()
+        } else {
+            panic!("key for {} not found", key.key)
+        }
+    }
+
+    /// Gets a value from the environment.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the value for the key is found, but has the wrong type.
+    pub fn try_get<'a, V: ValueType<'a>>(&'a self, key: Key<V>) -> Option<V> {
+        self.0
+            .map
+            .get(key.key)
+            .map(|value| value.to_inner_unchecked())
+    }
+
+    /// Adds a key/value, acting like a builder.
+    pub fn adding<'a, V: ValueType<'a>>(mut self, key: Key<V>, value: impl Into<V::Owned>) -> Env
+    where
+        V::Owned: Into<Value>,
+    {
+        let env = Arc::make_mut(&mut self.0);
+        env.map.insert(key.into(), value.into().into());
+        self
+    }
+
+    /// Sets a value in an environment.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the environment already has a value for the key, but it is
+    /// of a different type.
+    pub fn set<'a, V: ValueType<'a>>(&mut self, key: Key<V>, value: impl Into<V::Owned>)
+    where
+        V::Owned: Into<Value>,
+    {
+        let env = Arc::make_mut(&mut self.0);
+        let value = value.into().into();
+        let key = key.into();
+        // TODO: use of Entry might be more efficient
+        if let Some(existing) = env.map.get(&key) {
+            if !existing.is_same_type(&value) {
+                panic!(
+                    "Invalid type for key '{}': {:?} differs in kind from {:?}",
+                    key, existing, value
+                );
+            }
+        }
+        env.map.insert(key, value);
+    }
+}
+
+impl Debug for Value {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self {
+            Value::Point(p) => write!(f, "Point {:?}", p),
+            Value::Size(s) => write!(f, "Size {:?}", s),
+            Value::Rect(r) => write!(f, "Rect {:?}", r),
+            Value::Color(c) => write!(f, "Color {:?}", c),
+            Value::Float(x) => write!(f, "Float {}", x),
+            Value::String(s) => write!(f, "String {:?}", s),
+        }
+    }
+}
+
+impl<T> Key<T> {
+    pub const fn new(key: &'static str) -> Self {
```

Comment:
For docs:

```rust
/// Create a new strongly typed `Key` with the given string value.
/// The type of the key will be inferred.
/// 
/// # Examples
///
/// ```
/// use druid::env::Key;
/// use druid::piet::Color;
///
/// let float_key: Key<Float> = Key::new("a.very.good.float");
/// let color_key: Key<Color> = Key::new("a.very.nice.color");
/// ```
```

Reply:
+1 (I had to clean it up very slightly to compile)